src/pd/async/util.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::sync::RwLock;
+use std::sync::TryLockError;
+use std::time::Instant;
+use std::time::Duration;
+use std::thread;
+
+use futures::Poll;
+use futures::Async;
+use futures::Future;
+use futures::future::ok;
+use futures::future::{self, loop_fn, Loop};
+
+use kvproto::pdpb::GetMembersResponse;
+use kvproto::pdpb_grpc::PDAsync;
+use kvproto::pdpb_grpc::PDAsyncClient;
+
+use util::HandyRwLock;
+
+use super::super::PdFuture;
+use super::super::Result;
+use super::super::Error;
+use super::client::try_connect_leader;
+
+struct Bundle {
+    client: Arc<PDAsyncClient>,
+    members: GetMembersResponse,
+}
+
+/// Get a leader client asynchronous.
+pub struct LeaderClient {
+    inner: Arc<RwLock<Bundle>>,
+}
+
+impl LeaderClient {
+    pub fn new(client: PDAsyncClient, members: GetMembersResponse) -> LeaderClient {
+        LeaderClient {
+            inner: Arc::new(RwLock::new(Bundle {
+                client: Arc::new(client),
+                members: members.clone(),
+            })),
+        }
+    }
+
+    pub fn client<Req, Resp, G>(&self, retry: usize, req: Req, f: G) -> GetClient<Req, Resp, G>
+        where Req: Clone + 'static,
+              G: FnMut(Arc<PDAsyncClient>, Req) -> PdFuture<Resp> + Send + 'static
+    {
+        GetClient {
+            retry_count: retry,
+            bundle: self.inner.clone(),
+            req: req,
+            resp: None,
+            func: f,
+        }
+    }
+
+    pub fn get_client(&self) -> Arc<PDAsyncClient> {
+        self.inner.rl().client.clone()
+    }
+
+    pub fn set_client(&self, client: PDAsyncClient) {
+        let mut bundle = self.inner.wl();
+        bundle.client = Arc::new(client);
+    }
+
+    pub fn get_members(&self) -> GetMembersResponse {
+        self.inner.rl().members.clone()
+    }
+
+    pub fn set_members(&self, members: GetMembersResponse) {
+        let mut inner = self.inner.wl();
+        inner.members = members;
+    }
+}
+
+pub struct GetClient<Req, Resp, F> {
```

Comment:
What does Get mean here? Is the client only for get request? Does it have anything to do with the HTTP Verb Get?